pkg/apis/core/v1/defaults.go

```diff
@@ -133,32 +132,32 @@ func SetDefaults_Service(obj *v1.Service) {
 		obj.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeCluster
 	}
 
-	// if dualstack feature gate is on then we need to default
-	// Spec.IPFamily correctly. This is to cover the case
-	// when an existing cluster have been converted to dualstack
-	// i.e. it already contain services with Spec.IPFamily==nil
-	if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) &&
-		obj.Spec.Type != v1.ServiceTypeExternalName &&
-		obj.Spec.ClusterIP != "" && /*has an ip already set*/
-		obj.Spec.ClusterIP != "None" && /* not converting from ExternalName to other */
-		obj.Spec.IPFamily == nil /* family was not previously set */ {
+	// default obj.Spec.IPFamilyPolicy
+	if obj.Spec.Type != v1.ServiceTypeExternalName && obj.Spec.IPFamilyPolicy == nil {
+		// for single stack cases
+		// len(obj.Spec.ClusterIPs) == 0 || 1
+		// len(obj.Spec.IPFamilies) == 0 || 1
+		// we don't set it here (only during reserve/alloc)
+		// to allow web hooks to set it if needed
 
-		// there is a change that the ClusterIP (set by user) is unparsable.
-		// in this case, the family will be set mistakenly to ipv4 (because
-		// the util function does not parse errors *sigh*). The error
-		// will be caught in validation which asserts the validity of the
-		// IP and the service object will not be persisted with the wrong IP
-		// family
+		if len(obj.Spec.ClusterIPs) == 2 || len(obj.Spec.IPFamilies) == 2 {
+			requireDualStack := v1.RequireDualStack
+			obj.Spec.IPFamilyPolicy = &requireDualStack
+		}
+	}
 
-		ipv6 := v1.IPv6Protocol
-		ipv4 := v1.IPv4Protocol
-		if utilnet.IsIPv6String(obj.Spec.ClusterIP) {
-			obj.Spec.IPFamily = &ipv6
+	// the only allowed defaulting for RequiredDualStack
+	// if one family provided, we add the other
+	if obj.Spec.IPFamilyPolicy != nil && *(obj.Spec.IPFamilyPolicy) == v1.RequireDualStack && len(obj.Spec.IPFamilies) == 1 {
```

Comment:
empty IPFamilies are defaulted in allocation. I refrained from doing `.IPFamilies[n] = ipfamily(ClusterIP[n])` specifically because there is no way to return errors. So the entire `default family based on ips provided` is done in allocation where we can produce proper validation errors and return to user.

I will mark this as resolved. Please re-open if you want to readdress